ci/upload_test_to_db.py

```diff
@@ -0,0 +1,488 @@
+#!/usr/bin/env python3
+"""
+Upload pytest results to MySQL.
+Tables:
+ - ci_runs:    one row per run
+ - ci_tests:   one row per unique test
+ - ci_results: one row per test per run
+"""
+
+from __future__ import annotations
+
+import os
+import re
+import ast
+import json
+from functools import lru_cache
+from pathlib import Path
+from datetime import datetime, timezone
+from typing import Iterable, List, Tuple, Optional, Dict
+
+# pylint: disable=import-error
+import mysql.connector
+from mysql.connector import Error as MySQLError
+
+# -----------------------------
+# Constants
+# -----------------------------
+TEXT_LIMIT = 250
+BATCH_SIZE = 2000
+SKIP_FILES = {
+    "final_compiled_report.json",
+    "final_compiled_report.html",
+    "collect_module_log.jsonl",
+    "test_abort_log.json",
+    "test_abort_log.html",
+    "test_seg_fault_log.json",
+    "test_seg_fault_log.html",
+}
+DEFAULT_LABEL = "Skipped Upstream"
+
+
+# -----------------------------
+# Helpers
+# -----------------------------
+def shorten(s: Optional[str], limit: int = TEXT_LIMIT) -> Optional[str]:
+    """Return string limited to `limit` characters, or None."""
+    if s is None:
+        return None
+    s = str(s)
+    return s if len(s) <= limit else s[:limit]
+
+
+def extract_skip_reason(reason: str) -> str:
+    """Parse pytest skip longrepr tuple-string into its reason text.
+
+    Example input: "('/path/test_x.py', 42, 'Skipped: some reason')"
+    """
+    try:
+        tup = ast.literal_eval(reason)
```

Comment:
Is there a way to get skip reasons without `ast.literal_eval`? In my experience, evals like this cause problems down the line because 1) you can eval something nasty and 2) it's hard to understand what gets evaled when you want to debug. The `categorize_reasons` function and its rules mechanism seems liked it'd be powerful enough to handle the reason text plus some extra bits.